keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -1,1618 +1,1618 @@
+ surface the latency data (as this no longer fits the structure of a pod
+ condition). This data cannot be consumed by other controllers as mentioned in
+ User Stories section.
+ 
+ ### Report sandbox creation latency as an aggregated metric
+ 
+ The duration it took pod sandbox to become ready can be directly reported as a
+ prometheus metrics aggregated in a histogram. However, aggregating the data at
+ the Kubelet level prevents a metric collection service from classifying the data
+ based on interesting fields on a pod (runtime class, storage class of PVCs,
+ number of PVCs, etc) or using custom labels and annotations on pods that
+ indicate workload characteristics (that the cluster operator may wish to use as
+ a basis for aggregating the metrics).
+ 
+ This also prevents other controllers from acting on sandbox status as
+ mentioned in User Stories section.
+ 
+ ### Report sandbox creation stages using Kubelet tracing
+ 
+ The Kubelet is being instrumented to emit traces based on OpenTelemetry around
+ sandbox creation stages (as well several other parts of the pod lifecycle).
+ 
+ To implement the pod sandbox creation latency SLI/SLO use cases, the tracing
+ infrastructure needs to be able to:
+ - Collect all traces around CRI sandbox creation for all pods with no sampling.
+ - Look-up pod fields from API server (associated with a pod's trace) like
+ labels/annotations/storage classes of PVCs referred by the pod/runtimeclass/etc.
+ that is of interest to cluster operators and their users for classifying and
+ aggregating the metrics.
+ - Look-up a pod's Scheduled condition fields to determine the beginning of pod
+ sandbox creation.
+ 
+ Since the lookup of the pod fields and existing conditions is necessary for SLIs
+ around pod sandbox creation latency, surfacing the `PodHasNetwork` condition in
+ pod status will allow a metric collection service to directly access the
+ relevant data without requiring the ability to collect and parse OpenTelemetry
+ traces. As mentioned in the User Stories, popular community managed services
+ like Kube State Metrics can consume the `PodHasNetwork` condition with a trivial
+ set of changes. Enhancing them to collect and parse OpenTelemetry traces with no
+ sampling and mapping the data to associated data from API server data will be
+ complex from an engineering and operational perspective.
+ 
+ For controllers using the pod sandbox conditions to determine reconciliation
+ strategy, access to the pod is typically necessary while collecting and parsing
+ traces would be unusual.
+ 
+ ### Have CSI/CNI/CRI plugins mark their start and completion timestamps while setting up their respective portions for a pod
+ 
+ Each infrastructural plugin that Kubelet calls out to (in the process of setting
+ up a pod sandbox) can mark start and completion timestamps on the pod as
+ conditions. This approach would be similar to how readiness gates work today.
```

Comment:
I'm not quite buying this - if kubelet is the internal "switchboard" for running messages between plugins, that feels like a purely internal thing - don't make these plugins THEMSELVES update the "pre-readiness gates", make kubelet do it.